File: api/src/main/java/net/jqwik/api/Shrinkable.java
package net.jqwik.api;

import java.util.*;
import java.util.function.*;
import java.util.stream.*;

import org.apiguardian.api.*;
import org.jspecify.annotations.*;

import static org.apiguardian.api.API.Status.*;

@API(status = STABLE, since = "1.0")
public interface Shrinkable<T extends @Nullable Object> extends Comparable<Shrinkable<T>> {

	@API(status = INTERNAL)
	abstract class ShrinkableFacade {
		private static final ShrinkableFacade implementation;

		static {
			implementation = FacadeLoader.load(ShrinkableFacade.class);
		}

		public abstract <T> Shrinkable<T> unshrinkable(@Nullable Supplier<T> valueSupplier, ShrinkingDistance distance);

		public abstract <T, U> Shrinkable<U> map(Shrinkable<T> self, Function<T, U> mapper);

		public abstract <T> Shrinkable<T> filter(Shrinkable<T> self, Predicate<T> filter);

		public abstract <T, U> Shrinkable<U> flatMap(Shrinkable<T> self, Function<T, Arbitrary<U>> flatMapper, int tries, long randomSeed);
	}

	static <T> Shrinkable<T> unshrinkable(@Nullable T value) {
		return unshrinkable(value, ShrinkingDistance.MIN);
	}

	static <T> Shrinkable<T> unshrinkable(@Nullable T value, ShrinkingDistance distance) {
		return ShrinkableFacade.implementation.unshrinkable(() -> value, distance);
	}

	@API(status = INTERNAL)
	static <T> Shrinkable<T> supplyUnshrinkable(Supplier<T> supplier) {
		return ShrinkableFacade.implementation.unshrinkable(supplier, ShrinkingDistance.MIN);
	}

	/**
	 * Create value freshly, so that in case of mutable objects shrinking (and reporting)
	 * can rely on untouched values.
	 *
	 * @return An un-changed instance of the value represented by this shrinkable
	 */
	T value();

	/**
	 * Create a new and finite stream of smaller or same size shrinkables; size is measured by {@linkplain #distance()}.
	 * <p>
	 * Same size shrinkables are allowed but they have to iterate towards a single value to prevent endless shrinking.
	 * This also means that a shrinkable must never be in its own shrink stream!
	 *
	 * @return a finite stream of shrinking options
	 */
	@API(status = INTERNAL, since = "1.3.3")
	Stream<Shrinkable<T>> shrink();

	/**
	 * To be able to "move" values towards the end of collections while keeping some constraint constant
	 * it's necessary to grow a shrinkable by what another has been shrunk.
	 * One example is keeping a sum of values and still shrinking to the same resulting list.
	 *
	 * @param before The other shrinkable before shrinking
	 * @param after The other shrinkable after shrinking
	 * @return this shrinkable grown by the difference of before and after
	 */
	@API(status = INTERNAL, since = "1.3.3")
	default Optional<Shrinkable<T>> grow(Shrinkable<?> before, Shrinkable<?> after) {
		return Optional.empty();
	}

	/**
	 * Grow a shrinkable to allow broader searching in flat mapped shrinkables
	 *
	 * @return a finite stream of grown values
	 */
	@API(status = INTERNAL, since = "1.3.3")
	default Stream<Shrinkable<T>> grow() {
		return Stream.empty();
	}

	ShrinkingDistance distance();

	/**
	 * Sometimes simplifies test writing
	 *
	 * @return generic version of a shrinkable
	 */
	@SuppressWarnings("unchecked")
	@API(status = INTERNAL, since = "1.2.4")
	default Shrinkable<Object> asGeneric() {
		return (Shrinkable<Object>) this;
	}

	default <U> Shrinkable<U> map(Function<T, U> mapper) {
		return ShrinkableFacade.implementation.map(this, mapper);
	}

	default Shrinkable<T> filter(Predicate<T> filter) {
		return ShrinkableFacade.implementation.filter(this, filter);
	}

	default <U> Shrinkable<U> flatMap(Function<T, Arbitrary<U>> flatMapper, int tries, long randomSeed) {
		return ShrinkableFacade.implementation.flatMap(this, flatMapper, tries, randomSeed);
	}

	@SuppressWarnings("unchecked")
	@Override
	@API(status = INTERNAL)
	default int compareTo(Shrinkable<T> other) {
		int comparison = this.distance().compareTo(other.distance());
		if (comparison == 0) {
			T value = value();
			if (value instanceof Comparable && this.getClass().equals(other.getClass())) {
				return ((Comparable<T>) value).compareTo(other.value());
			}
		}
		return comparison;
	}

	@API(status = INTERNAL)
	default Shrinkable<T> makeUnshrinkable() {
		return ShrinkableFacade.implementation.unshrinkable(this::value, this.distance());
	}

}

File: api/src/main/kotlin/net/jqwik/api/Shrinkable.kt
package net.jqwik.api

import org.apiguardian.api.API
import java.util.*
import java.util.function.Function
import java.util.function.Predicate
import java.util.function.Supplier
import java.util.stream.Stream

@API(status = API.Status.STABLE, since = "1.0")
interface Shrinkable<out T> : Comparable<Shrinkable<@UnsafeVariance T>> {
    @API(status = API.Status.INTERNAL)
    abstract class ShrinkableFacade {
        abstract fun <T> unshrinkable(valueSupplier: Supplier<T>, distance: ShrinkingDistance): Shrinkable<T>
        abstract fun <T, U> map(self: Shrinkable<T>, mapper: Function<in T, out U>): Shrinkable<U>
        abstract fun <T> filter(self: Shrinkable<T>, filter: Predicate<in T>): Shrinkable<T>
        abstract fun <T, U> flatMap(
            self: Shrinkable<T>,
            flatMapper: Function<in T, Arbitrary<U>?>?,
            tries: Int,
            randomSeed: Long
        ): Shrinkable<U>?

        companion object {
            val implementation = FacadeLoader.load(
                ShrinkableFacade::class.java
            )
        }
    }

    /**
     * Create value freshly, so that in case of mutable objects shrinking (and reporting)
     * can rely on untouched values.
     *
     * @return An un-changed instance of the value represented by this shrinkable
     */
    fun value(): T

    /**
     * Create a new and finite stream of smaller or same size shrinkables; size is measured by [.distance].
     *
     *
     * Same size shrinkables are allowed but they have to iterate towards a single value to prevent endless shrinking.
     * This also means that a shrinkable must never be in its own shrink stream!
     *
     * @return a finite stream of shrinking options
     */
    @API(status = API.Status.INTERNAL, since = "1.3.3")
    fun shrink(): Stream<out Shrinkable<T>>

    /**
     * To be able to "move" values towards the end of collections while keeping some constraint constant
     * it's necessary to grow a shrinkable by what another has been shrunk.
     * One example is keeping a sum of values and still shrinking to the same resulting list.
     *
     * @param before The other shrinkable before shrinking
     * @param after The other shrinkable after shrinking
     * @return this shrinkable grown by the difference of before and after
     */
    @API(status = API.Status.INTERNAL, since = "1.3.3")
    fun grow(before: Shrinkable<*>?, after: Shrinkable<*>?): Optional<out Shrinkable<T>> {
        return Optional.empty()
    }

    /**
     * Grow a shrinkable to allow broader searching in flat mapped shrinkables
     *
     * @return a finite stream of grown values
     */
    @API(status = API.Status.INTERNAL, since = "1.3.3")
    fun grow(): Stream<out Shrinkable<T>>? {
        return Stream.empty()
    }

    fun distance(): ShrinkingDistance

    /**
     * Sometimes simplifies test writing
     *
     * @return generic version of a shrinkable
     */
    @API(status = API.Status.INTERNAL, since = "1.2.4")
    fun asGeneric(): Shrinkable<Any?>? {
        return this as Shrinkable<Any?>
    }

    fun <U> map(mapper: Function<in T, U>): Shrinkable<U> {
        return ShrinkableFacade.implementation!!.map(this, mapper)
    }

    fun filter(filter: Predicate<in T>): Shrinkable<T> {
        return ShrinkableFacade.implementation!!.filter(this, filter)
    }

    fun <U> flatMap(flatMapper: Function<in T, Arbitrary<U>?>?, tries: Int, randomSeed: Long): Shrinkable<U>? {
        return ShrinkableFacade.implementation!!.flatMap(this, flatMapper, tries, randomSeed)
    }

    @API(status = API.Status.INTERNAL)
    override fun compareTo(other: Shrinkable<@UnsafeVariance T>): Int {
        val comparison = distance().compareTo(other.distance())
        if (comparison == 0) {
            val value = value()
            if (value is Comparable<*> && this.javaClass == other.javaClass) {
                return (value as Comparable<T>).compareTo(other.value())
            }
        }
        return comparison
    }

    @API(status = API.Status.INTERNAL)
    fun makeUnshrinkable(): Shrinkable<T> {
        return ShrinkableFacade.implementation!!.unshrinkable({ value() }, distance())
    }

    companion object {
        @JvmStatic
        fun <T> unshrinkable(value: T): Shrinkable<T> {
            return unshrinkable<T>(value, ShrinkingDistance.of(0))
        }

        @JvmStatic
		fun <T> unshrinkable(value: T, distance: ShrinkingDistance): Shrinkable<T> {
            return ShrinkableFacade.implementation!!.unshrinkable({ value }, distance)
        }

        @JvmStatic
		@API(status = API.Status.INTERNAL)
        fun <T> supplyUnshrinkable(supplier: Supplier<T>, cacheImmutables: Boolean): Shrinkable<T> {
            return ShrinkableFacade.implementation!!.unshrinkable(supplier, ShrinkingDistance.of(0))
        }
    }
}